arangod/Pregel/Actor/include/Actor/Message.h

```diff
@@ -0,0 +1,134 @@
+////////////////////////////////////////////////////////////////////////////////
+/// DISCLAIMER
+///
+/// Copyright 2014-2022 ArangoDB GmbH, Cologne, Germany
+/// Copyright 2004-2014 triAGENS GmbH, Cologne, Germany
+///
+/// Licensed under the Apache License, Version 2.0 (the "License");
+/// you may not use this file except in compliance with the License.
+/// You may obtain a copy of the License at
+///
+///     http://www.apache.org/licenses/LICENSE-2.0
+///
+/// Unless required by applicable law or agreed to in writing, software
+/// distributed under the License is distributed on an "AS IS" BASIS,
+/// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+/// See the License for the specific language governing permissions and
+/// limitations under the License.
+///
+/// Copyright holder is ArangoDB GmbH, Cologne, Germany
+///
+/// @author Julia Volmer
+/// @author Markus Pfeiffer
+////////////////////////////////////////////////////////////////////////////////
+#pragma once
+
+#include <Inspection/VPackWithErrorT.h>
+#include <memory>
+#include <variant>
+
+#include <velocypack/Builder.h>
+
+#include "ActorPID.h"
+#include "MPSCQueue.h"
+
+namespace {
+// helper type for the visitor
+template<class... Ts>
+struct overloaded : Ts... {
+  using Ts::operator()...;
+};
+template<class... Ts>
+overloaded(Ts...) -> overloaded<Ts...>;
+}  // namespace
+
+namespace arangodb::pregel::actor {
+
+struct MessagePayloadBase {
+  virtual ~MessagePayloadBase() = default;
+};
+
+template<typename Payload>
+struct MessagePayload : MessagePayloadBase {
```

Comment:
Maybe? In such case it will be single allocation instead of two for process
```suggestion
struct MessagePayload : MessagePayloadBase, Node {
```

Reply:
I'm not sure I understand why this removes one allocation, which two allocations do you mean?

Reply:
Now you need one allocation for Message another for payload

Reply:
I still don't understand how we can get rid of one allocation:
In Runtime::dispatchLocally, we create the MessagePayload (I guess that is the first allocation you refer to). In the upper Actor::push we create an InternalMessage out of this message payload (I guess that is the second allocation). If I understand your correctly, you say that we can get rid of the second allocation if MessagePayload inherits from Node. 
Internal Message inherits from Node and the Queue we use in the Actor is a Queue with a Node pointing at an InternalMessage type. InternalMessage consists not only of the MessagePayload, but also of a sender, so how can we get rid of the second allocation when MessagePayload is a Node?

Reply:
> In Runtime::dispatchLocally, we create the MessagePayload (I guess that is the first allocation you refer to).

Now it's (external message) created on stack (if I see all usages), so it's ok for me. 
You fixed it in some commit when you fixed other comment about `unique_ptr`.
So I resolve it.

> If I understand your correctly, you say that we can get rid of the second allocation if MessagePayload inherits from Node.

Yes, comment was about that.

> InternalMessage consists not only of the MessagePayload, but also of a sender, so how can we get rid of the second allocation when MessagePayload is a Node?

Idea was Just protected inherit ExternalMessage from InternalMessage.
So when user allocate ExternalMessage he is also allocate InternalMessage.
So in `proccess(..., std::unique_ptr<ExternalMessage> p)` you can access to the InternalMessage and don't need to allocate it, only change maybe.